### PR TITLE
Turn save-fuel-price-data into a task we can run in AWS

### PR DESCRIPTION
**Pull request by chrisgilmerproj**
## Description

Requires we first set up AWS in https://github.com/transcom/ppp-infra/pull/487

This PR containerizes `bin/save-fuel-price-data` into something we can deploy to AWS ECS Fargate as a scheduled task.

## Reviewer Notes

I've made a few updates to the script so we can target the correct database parameters.

## Setup

You can immediately try the containerized version using:

```sh
make tasks_save_fuel_price_data
```

This will build a container locally and then run the binary, which is the only thing in that container other than the DB definition.  All environment vars are passed through to the container but we could instead use flags if needed.

The rest of the deploy process is a bit more hairy and follows this order:

1. Build the binary
1. Build the Dockerfile with the new binary to make a new image
1. Push the new image to our ECR repository in AWS
1. Run a script to deploy that image to ECS Fargate using a Task Definition

The script needs to be rewritten and tested as its only a copy of `bin/ecs-deploy-service-container` that has a couple changes to target scheduled tasks.  Generally the script will do this:

1. Render a new Task Definition which is how ECS Fargate uses to update/deploy the container
1. Register the new Task Definition
1. Update the Scheduled Task service (using green/blue deploy)

## Download the container to check it out:

```
exec $(aws ecr get-login --no-include-email)
docker pull <image>
```

## Open Questions

- Do we deploy tasks concurrently with the app or do we do it only on successful app deployment?
- What does rollback mean for this task?
- Are we going to have issues making a secure connection between the binary and the RDS instance?

## Code Review Verification Steps

* [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
* [ ] Request review from a member of a different team.
* [ ] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/163909133) for this change

### CODE REVIEW

**Comment by chrisgilmerproj** [comment]
This worked! https://circleci.com/gh/transcom/mymove/94932

I saw a successful deploy of a new image and the updated task.

**Comment by chrisgilmerproj** [comment]
@kimallen @reggieriser @pjdufour-truss @rdhariwal - This is ready for review! 

I'd love for a closer look at the two scripts I've created/modified. The circleci config may need scrutiny too. And finally I am hoping to get in #2049 before this one so I can eliminate a lot of the changes I made to the Makefile in this PR.

File: Dockerfile.tasks [hunk 1]
```diff
@@ -0,0 +1,8 @@
+FROM gcr.io/distroless/base
+
+COPY bin/rds-combined-ca-bundle.pem /bin/rds-combined-ca-bundle.pem
+COPY bin/chamber /bin/chamber
+COPY bin/save-fuel-price-data /bin/save-fuel-price-data
+COPY config/database.yml /config/database.yml
```
Comment (chrisgilmerproj) [inline Dockerfile.tasks]: Remove this.

File: Dockerfile.tasks_local [hunk 2]
```diff
@@ -0,0 +1,6 @@
+FROM gcr.io/distroless/base
+
+COPY bin_linux/save-fuel-price-data /bin/save-fuel-price-data
+COPY config/database.yml /config/database.yml
```
Comment (chrisgilmerproj) [inline Dockerfile.tasks_local]: Remove this.

File: cmd/save_fuel_price_data/main.go [hunk 3]
```diff
@@ -6,46 +6,90 @@ import (
...
 	"github.com/spf13/pflag"
 	"github.com/spf13/viper"
+	"go.uber.org/zap"
 
+	"github.com/transcom/mymove/pkg/cli"
 	"github.com/transcom/mymove/pkg/logging"
 	"github.com/transcom/mymove/pkg/services/fuelprice"
 )
 
+func checkConfig(v *viper.Viper, logger logger) error {
```
Comment (kimallen) [inline cmd/save_fuel_price_data/main.go]: I ran the CLI tool, and got this error for this line `cmd/save_fuel_price_data/main.go:18:41: undefined: logger`

File: Dockerfile [hunk 4]
```diff
@@ -1,8 +1,8 @@
 FROM gcr.io/distroless/base
 
 COPY bin/rds-combined-ca-bundle.pem /bin/rds-combined-ca-bundle.pem
-COPY bin/milmove /bin/milmove
 COPY bin/chamber /bin/chamber
+COPY bin/milmove /bin/milmove
```
Comment (rdhariwal) [inline Dockerfile]: does the order matter?

Comment (chrisgilmerproj) [reply Dockerfile]: Not really. I was tidying things up and probably did a `vimdiff` and made it look like another `Dockerfile`.

File: cmd/save_fuel_price_data/main.go (hunk 3 above)
Comment (chrisgilmerproj) [reply cmd/save_fuel_price_data/main.go]: It builds fine with `make bin/save-fuel-price-data` but doesn’t work with `go run`. This is apparently a problem in other places too so I'll figure out how to manage it.

File: .circleci/config.yml [hunk 5]
```diff
@@ -67,6 +67,20 @@ commands:
...
+      - checkout
+      - setup_remote_docker:
+          docker_layer_caching: true
+      - restore_cache:
+          keys:
+            - go-mod-souces-v1-{{ checksum "go.sum" }}
+      - run: make bin/ecs-deploy-task-container
+      - deploy:
+          name: Deploy task service
+          command: scripts/do-exclusively --job-name ${CIRCLE_JOB} bin/ecs-deploy-task-container --aws-account-id ${AWS_ACCOUNT_ID} --aws-region ${AWS_DEFAULT_REGION} --service app --environment experimental --repository-name app-tasks --image-tag git-${CIRCLE_SHA1} --rule save-fuel-price-data
```
Comment (chrisgilmerproj) [inline .circleci/config.yml]: change based on environment

File: cmd/ecs-deploy-task-container/main.go [hunk 6]
```diff
@@ -0,0 +1,541 @@
...
+
+	// Get the current task definition
+	currentTaskDefArnStr := *currentTarget.EcsParameters.TaskDefinitionArn
+	logger.Println(fmt.Sprintf("Current Task Def Arn: %s", currentTaskDefArnStr))
+	currentTaskDefArn, err := arn.Parse(currentTaskDefArnStr)
+	if err != nil {
+		quit(logger, nil, errors.Wrap(err, "Unable to parse current task definition arn"))
+	}
+
+	// TODO: This isn't clean and could probably use a regex
```
Comment (chrisgilmerproj) [inline cmd/ecs-deploy-task-container/main.go]: fix todo

File: cmd/ecs-deploy-task-container/main.go [hunk 7]
```diff
@@ -0,0 +1,541 @@
...
+		},
+		RegistryId:     aws.String(registryID),
+		RepositoryName: aws.String(repositoryName),
+	})
+	if err != nil {
+		quit(logger, nil, errors.Wrapf(err, "unable retrieving image from %s", imageName))
+	}
+
+	// Get the database host using the instance identifier
+	// TODO: Allow passing in from DB_HOST
```
Comment (chrisgilmerproj) [inline cmd/ecs-deploy-task-container/main.go]: fix todo

File: cmd/save_fuel_price_data/main.go (hunk 3 above)
Comment (chrisgilmerproj) [reply cmd/save_fuel_price_data/main.go]: Turns out you have to pass it either all the files or the name of the package. So this works:

```
go run github.com/transcom/mymove/cmd/save_fuel_price_data
```

Or this:

```
go run cmd/save_fuel_price_data/main.go cmd/save_fuel_price_data/logger.go
```

File: cmd/save_fuel_price_data/main.go [hunk 8]
```diff
@@ -6,46 +6,90 @@ import (
...
+	cli.InitEIAFlags(flag)
+
+	// Verbose
+	cli.InitVerboseFlags(flag)
+
+	// Don't sort flags
+	flag.SortFlags = false
+}
+
 // Command: go run cmd/save_fuel_price_data/main.go
```
Comment (kimallen) [inline cmd/save_fuel_price_data/main.go]: Now that you figured out the proper command to run it, will you change this comment accordingly?

**Comment by chrisgilmerproj** [comment]
> `save_fuel_price_data` looks good to me- ran fine in development. Just a comment to edit the comment for the actual command to use on CLI.

@kimallen - I made the changes. Can you approve? I'll still wait for one approval from Infra too, but I'd like at least your sign off.

**Comment by chrisgilmerproj** [comment]
@rdhariwal - if you have no more comments can you approve?

**Pull request merged**